[자동차 경주 리팩터링] 팡태 미션 제출합니다

Hevton opened the pull request:
안녕하세요 리뷰어님! 안드로이드 2조 팡태입니다.
안내해주신 가이드 모두 확인했습니다! step2 미션도 잘 부탁드립니다. 감기가 유행인데 감기 조심하세요.

한 가지 이야기해보고 싶은 점이 있습니다.
Car의 움직임 정보를 출력할때, 현재 코드에서는
`OutputView`의 `printPhase()` 함수에서 `'자동차이름 : 횟수'`  형식을 만들어서 출력하고 있습니다.
```
// OutputView.kt
    fun printPhase(cars: Cars) {
        cars.cars.forEach {
            println("${it.name} :" + "-".repeat(it.position))
        }
        println()
    }

```
이는 View가 **출력 형식을 지정**해준다는 점에서 납득이 간다고 생각하고 있습니다.

그치만 Car의 정보를 출력하는 만큼, Car 클래스의 toString() 을 재정의하여 포맷을 만드는
```
// Car.kt
    override
    fun toString(): String {
        val scoreLine = "-".repeat(position)
        return "$name : $scoreLine\n"
    }
```
이 함수를 OutputView에서 가져다 쓰는 것도 Car의 책임을 Car에서 다룬다는 면에서 적절하지 않나 생각합니다.

물론 정답은 없을 수 있다고 생각하고 있습니다.
정답이라는걸 알고 싶어서 보다는, 저는 보지 못하는 리뷰어님의 시야를 알고 싶어서 질문 남기게 되었습니다.

Commit:
style: 코드 스타일을 ktlint에 맞게 조정

Commit:
refactor(TryCount): 상수 선언 위치 변경

- 이전: TryCount 밖에 상수로 선언했었음
- 이후: TryCount companion object 안에 선언. 결국 TryCount 에서만 쓰이는 상수이므로 구조적으로 더 나은 방향이라고 생각되었다

Commit:
refactor(Cars): 상수 선언 위치 변경

- 이전: Cars 밖에 상수로 선언했었음
- 이후: Cars companion object 안에 선언. 결국 Cars 에서만 쓰이는 상수이므로 구조적으로 더 나은 방향이라고 생각되었다

Commit:
refactor(OutputView): 상수 접근 제어자 변경

OutputView 에서만 쓰이는 상수이기에, private 접근 제어자 명시

Commit:
refactor(InputView): 상수 접근 제어자 변경

InputView 에서만 쓰이는 상수이기에, private 접근 제어자 명시

Commit:
refactor(RacingGame): 함수 호출 코드를 한 줄로 압축

Commit:
refactor(Car): 숫자 상수화 적용

Gyuil-Hwnag commented on build.gradle.kts:
```diff
@@ -16,7 +16,7 @@ dependencies {
     testImplementation("org.junit.jupiter", "junit-jupiter", "5.10.2")
     testImplementation("org.assertj", "assertj-core", "3.25.3")
     testImplementation("io.kotest", "kotest-runner-junit5", "5.8.0")
-    testImplementation ("org.junit.jupiter:junit-jupiter-params:5.10.2")
+    testImplementation("org.junit.jupiter:junit-jupiter-params:5.10.2")
 }
```
컨벤션을 맞춘다고 하면 해당 코드는 과연 컨벤션에 맞을까요??🤔

Hevton replied:
안녕하세요. 의존성 중복이 있다는 것을 확인했습니다.. ㅜㅜ
`./gradlew clean klintCheck` 를 진행했지만 문제로 인식받지 못한다는 것을 알게 되었습니다. 감사합니다.

Gyuil-Hwnag commented on src/main/kotlin/controller/RacingGame.kt:
```diff
@@ -24,17 +23,18 @@ class RacingGame(
 
     private fun getWinners(cars: Cars) {
-        val winners = cars.getWinners()
         outputView.printResultHeader()
-        outputView.printWinner(winners)
+        outputView.printWinner(cars.getWinners())
     }
```
getWinners() 함수의 기능을 좀 더 명확하게 하도록 수정하셨네요👍

Gyuil-Hwnag commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -1,13 +1,12 @@
-
-    var position: Int = 0
+    var position: Int = INIT_POSITION
         private set
```
외부 setter를 막기 위한 private set 키워드 좋습니다👍

Gyuil-Hwnag commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -1,13 +1,12 @@
     init {
         require(name.isNotBlank()) { EXCEPTION_CAR_NAME_IS_BLANK }
-        require(name.length <= 5) { EXCEPTION_INVALID_CAR_NAME_LENGTH }
+        require(name.length <= NAME_UPPER_BOUND) { EXCEPTION_INVALID_CAR_NAME_LENGTH }
```
조건에 해당하는 값들에 대한 상수화 좋네요~

Gyuil-Hwnag commented on src/main/kotlin/domain/NumberGenerator.kt:
```diff
@@ -2,4 +2,4 @@ package domain
     fun generate(): Int
-}
\ No newline at end of file
+}
```
EOF 경고 잘 없애 주셨군요~

Gyuil-Hwnag commented on src/main/kotlin/domain/Winners.kt:
```diff
@@ -1,3 +1,3 @@
 
-data class Winners(val winners: List<Car>)
\ No newline at end of file
+data class Winners(val winners: List<Car>)
```
Winners 데이터 클래스가 존재하는 이유에 대해서 궁금합니다!

Hevton replied:
처음에는 우승자를 List<Car> 타입 그대로 리턴할까 생각했지만
Winners라는 클래스로 감싸서 리턴을 하면, OutputView 클래스에서 우승자를 출력하는 등 우승자 데이터를 다룰 때 의미를 더 명확하게 인지할 수 있지 않을까 생각했습니다..

Gyuil-Hwnag replied:
추후 Winner에 대한 정보가 추가된다면 유용하게 사용될 수 있겠군요!👍

Hevton replied:
@Gyuil-Hwnag 네 현재는 추가 정보가 없으므로 없어도 될 것 같네요..!

Gyuil-Hwnag commented on src/main/kotlin/view/InputView.kt:
```diff
@@ -1,7 +1,6 @@
 package view
 
 class InputView {
```
InputView를 Class로 설정하시는 이유에 대해서 궁금합니다!🤔

Hevton replied:
말씀 감사합니다. InputView와 OutputView의 정의 기준을 미처 생각하지 않고 자연스레 클래스로 구현했습니다.
둘은 상태를 갖지도, 의존성을 주입받지도 않기에 object로 정의하는게 적절하다고 생각하게 되었습니다.

Gyuil-Hwnag commented on src/test/kotlin/domain/CarsTest.kt:
```diff
@@ -34,5 +34,4 @@ class CarsTest {
-
-}
\ No newline at end of file
+}
```
CarsTest를 살펴보았을 때 getWinners() 또한 중요한 주요 기능중 하나일텐데 우승자 체크에 대한 테스트 내용이 없는 것 같습니다!

Hevton replied:
감사합니다. 제가 놓치고 있었습니다.

Gyuil-Hwnag reviewed (changes requested):
팡테 안녕하세요!

자동자 경주 리팩토링 미션 구현 고생 많으셨습니다!
미션 구현을 너무 잘 해주셨습니다 💯

고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.
step2 너무 잘해주셔서 논의해보면 좋을 내용들이 몇개 있어서 해당 내용들 위주로 남겨두었습니다~

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 
화이팅입니다💪

Gyuil-Hwnag commented:
> 안녕하세요 리뷰어님! 안드로이드 2조 팡태입니다. 안내해주신 가이드 모두 확인했습니다! step2 미션도 잘 부탁드립니다. 감기가 유행인데 감기 조심하세요.
> 
> 한 가지 이야기해보고 싶은 점이 있습니다. Car의 움직임 정보를 출력할때, 현재 코드에서는 `OutputView`의 `printPhase()` 함수에서 `'자동차이름 : 횟수'` 형식을 만들어서 출력하고 있습니다.
> 
> ```
> // OutputView.kt
>     fun printPhase(cars: Cars) {
>         cars.cars.forEach {
>             println("${it.name} :" + "-".repeat(it.position))
>         }
>         println()
>     }
> ```
> 
> 이는 View가 **출력 형식을 지정**해준다는 점에서 납득이 간다고 생각하고 있습니다.
> 
> 그치만 Car의 정보를 출력하는 만큼, Car 클래스의 toString() 을 재정의하여 포맷을 만드는
> 
> ```
> // Car.kt
>     override
>     fun toString(): String {
>         val scoreLine = "-".repeat(position)
>         return "$name : $scoreLine\n"
>     }
> ```
> 
> 이 함수를 OutputView에서 가져다 쓰는 것도 Car의 책임을 Car에서 다룬다는 면에서 적절하지 않나 생각합니다.
> 
> 물론 정답은 없을 수 있다고 생각하고 있습니다. 정답이라는걸 알고 싶어서 보다는, 저는 보지 못하는 리뷰어님의 시야를 알고 싶어서 질문 남기게 되었습니다.

@Hevton 
팡태 안녕하세요~
해당 고민 정말 좋은 고민인 것 같습니다!!
흠~ Car.toString()이 언제나 "자동차이름 : 횟수" 만을 출력한다면 확실히 아래의 방법이 좋아 보이기는 합니다👍

Commit:
refactor(build): 의존성 중복 제거

Commit:
test(CarsTest): 우승자 확인 테스트 추가

Commit:
refactor(InputView): object로 변경

-이전: InputView가 Class로 정의되어 있다
-이후: InputView는 상태를 갖지도, 의존성을 주입받지도 않으므로 object로 변경

Commit:
refactor(OutputView): object로 변경

-이전: OutputView Class로 정의되어 있다
-이후: OutputView는 상태를 갖지도, 의존성을 주입받지도 않으므로 object로 변경

Gyuil-Hwnag commented on src/main/kotlin/view/InputView.kt:
```diff
@@ -1,6 +1,9 @@
 package view
 
-class InputView {
+object InputView {
+    private const val HEADER_READ_CARS = "경주할 자동차 이름을 입력하세요.(이름은 쉼표(,) 기준으로 구분)"
+    private const val INPUT_IS_NULL = "입력이 있어야 합니다"
+    private const val HEADER_READ_TRY_COUNT = "시도할 횟수는 몇 회인가요?"
 
```
InputView에 대해서 Object로 바꿔주셨군요👍

그렇다면 추가적으로 고민을 해보면 좋은 내용일 것 같아서 코멘트 남겨둡니다!!
Object 와 Class는 어떠한 차이점이 있을까요??
어떤 경우에 Object 를 사용하고 어떤 경우에 Class를 사용하면 좋을까요??🤔

Hevton replied:
InputView를 class가 아닌 object로 바꿔준 이유는,
클래스의 경우 여러 인스턴스가 있을 때 각각 다른 상태값들을 가질 경우 클래스로 구현한 의미가 있다고 생각합니다.
즉 InputView는 상태값을 갖지 않으며, 또한 의존 대상이 없고(=의존성을 주입받지 않고) 이에 더해 생성자의 필요성도 없기 때문이라고 생각했습니다.

Gyuil-Hwnag replied:
정확한 대답이었네요ㅎㅎ
타당한 이유 정말 좋습니다💯💯💯

Gyuil-Hwnag commented on src/main/kotlin/view/OutputView.kt:
```diff
@@ -16,13 +18,7 @@ class OutputView {
 
     fun printWinner(winners: Winners) {
         println(
-            "$HEADER_WINNERS " +
-                    "${winners.winners.joinToString(",") { it.name }} "
+            "$HEADER_WINNERS " + "${winners.winners.joinToString(",") { it.name }} ",
```
Cars 에서 고민해보았던 toString()을 Winner 객체에서도 toString()을 오버라이딩 할 수 있지 않을까요??🤔
반드시 수정이 필요한 내용은 아니고 고민해주셨던 Car의 toString() 오버라이딩 관련해서 생각나서 적어두었습니다~

Hevton replied:
맞는 말씀이십니다. 질문드렸던 내용에 이어 다음 미션을 진행할 때는 이러한 방식 대신 toString을 활용해보려고 합니다!

Gyuil-Hwnag commented on src/test/kotlin/domain/CarsTest.kt:
```diff
@@ -35,4 +36,10 @@ class CarsTest {
\ No newline at end of file
+    @Test
+    fun `결과를 기반으로 우승자 확인`() {
+        val cars = Cars.from("aa,bb,cc", ExplicitNumberGenerator(4))
```
반드시 수정이 필요한 내용은 아닙니다!!
4 라는 값에 대해서는 상수값으로 정의해두고 사용할 수도 있겠다는 생각이 듭니다💪
이동이 성공하였을 때를 테스트 할때는 MOVE_SUCCESS, MOVE_FAIL(예시)로 정의해두는 것도 좋아보일 것 같습니다!!

Hevton replied:
감사합니다 해당 부분 수정하겠습니다!

Gyuil-Hwnag commented on src/test/kotlin/domain/CarsTest.kt:
```diff
@@ -35,4 +36,10 @@ class CarsTest {
+        cars.startPhase()
+        assertEquals(listOf("aa", "bb", "cc"), cars.getWinners().winners.map { it.name })
+    }
+}
```
테스트 코드의 경우 로직을 검증하는 것이니 예외 케이스를 반영하여, 여러가지의 경우를 판별해보는 것도 좋을 것 같다는 생각이 들긴 합니다~

Gyuil-Hwnag commented on src/test/kotlin/domain/ExplicitNumberGenerator.kt:
```diff
@@ -1,5 +1,5 @@
-    override fun generate() =  input
-}
\ No newline at end of file
+class ExplicitNumberGenerator(private val input: Int) : NumberGenerator {
+    override fun generate() = input
+}
```
input 파라미터의 경우 불변 파라미터인데 private 키워드로 접근 지정자를 설정해주신 이유에 대해서 궁금합니다🤔

Hevton replied:
ExplicitNumberGenerator는 테스트에서만 쓰이는 클래스로, 테스트를 진행할 사용자가 인자로 값을 넘겨야 하는 만큼 사용자가 명시적으로 값을 알고 있습니다.
그러므로 이 클래스의 변수를 외부에서 접근할 경우가 없다고 생각해서 private 를 자연스레 붙였습니다.

Gyuil-Hwnag commented on src/main/kotlin/controller/RacingGame.kt:
```diff
@@ -20,21 +17,22 @@ class RacingGame(
-
-}
\ No newline at end of file
+}
```
EOF 경고 제거 굿입니다👍

Hevton replied:
@Gyuil-Hwnag 님 안녕하세요. 마지막 날 까지도 리뷰를 친히 도와주셔 감사합니다.
남겨주신 추가적인 말씀을 토대로 추가 구현을 진행했습니다.

질문 남겨드렸던 것 처럼, toString에 대한 활용을 앞으로의 미션에서 적용해보겠습니다.
또한 마지막으로 남겨주신 리뷰어님의 피드백을 기반으로 돌아보니, 모든 자동차의 일괄적인 움직임을 테스트하는 것은 가능하지만
각 자동차의 움직임에 차별점을 두어 테스트하는 것은 제 의존성 주입 방식 구성으로는 아직 부족하다고 느꼈습니다.

테스트코드를 작성할 때면 제 코드가 과연 잘 구성되어 있는지를 알아가게 되는 것 같습니다.
앞으로 이러한 점들을 기반으로 향후 미션에서 더 고민하고 적용할 수 있는 사람이 되겠습니다. 
1주차 미션 리뷰 도와주셔서 감사합니다.

Gyuil-Hwnag reviewed (changes requested):
팡태 안녕하세요~
해당 미션 너무 고생 많으셨습니다! 
리뷰 반영 정말 잘 해주셨네요.💯

도움이 될만한 코멘트 달았으니 확인해 보시면 좋을 것 같습니다!

너무 잘 반영 해주셔서 논의해보면 좋을 것 같은 내용들 남겨둔 내용들 입니다~
팡태의 판단하에 더 이상 수정될 것이 없다고 느껴지시면 알려주세요!!
알려주시면 해당 미션은 여기서 머지하도록 하겠습니다

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
다음 미션도 화이팅입니다💪

Commit:
refactor(CarsTest): 상수화 & 우승자 테스트 추가

Hevton commented:
@Gyuil-Hwnag 님 안녕하세요. 마지막 날 까지도 리뷰를 친히 도와주셔 감사합니다.
남겨주신 추가적인 말씀을 토대로 추가 구현을 진행했습니다.

질문 남겨드렸던 것 처럼, toString에 대한 활용을 앞으로의 미션에서 적용해보며 구현 차이를 느껴보겠습니다.
또한 마지막으로 남겨주신 리뷰어님의 피드백을 기반으로 돌아보니, 모든 자동차의 일괄적인 움직임을 기반으로 테스트하는 것은 가능하지만
각 자동차의 움직임에 차별점을 두어 테스트하는 것은 제가 구현한 의존성 주입 설계 구성으로는 아직 부족하다고 느꼈습니다.
물론 전자의 테스트만으로도 검증에 문제가 없으므로 후자가 꼭 필요한 테스트는 아니지만, 이런 생각까지 해볼 수 있게 되어 감사합니다.

앞으로 이러한 점들을 기반으로 향후 미션에서 더 넓게 생각하고 적용할 수 있는 사람이 되겠습니다.
1주차 미션 리뷰 도와주셔서 감사합니다.

Gyuil-Hwnag commented on src/test/kotlin/domain/CarsTest.kt:
```diff
@@ -23,16 +24,34 @@ class CarsTest {
+        assertEquals(listOf("aa", "bb", "cc"), cars.getWinners().winners.map { it.name })
+    }
+
+    companion object {
+        private const val MOVE = 4
+        private const val STOP = 3
+    }
+}
```
테스트 코드가 늘어날 경우 지정해놓은 상태값들을 통한 관리는 가독성 측면에서 좀 더 좋아지는 것 같습니다👍

Gyuil-Hwnag reviewed (approved):
안녕하세요 팡태~
해당 미션 너무 고생 많으셨습니다! 
리뷰 반영 정말 잘 해주셨네요.

해당 미션은 다음 미션 진행을 위하여 머지하도록 하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
다음 미션도 화이팅입니다💪

Gyuil-Hwnag commented:
> @Gyuil-Hwnag 님 안녕하세요. 마지막 날 까지도 리뷰를 친히 도와주셔 감사합니다. 남겨주신 추가적인 말씀을 토대로 추가 구현을 진행했습니다.
> 
> 질문 남겨드렸던 것 처럼, toString에 대한 활용을 앞으로의 미션에서 적용해보며 구현 차이를 느껴보겠습니다. 또한 마지막으로 남겨주신 리뷰어님의 피드백을 기반으로 돌아보니, 모든 자동차의 일괄적인 움직임을 기반으로 테스트하는 것은 가능하지만 각 자동차의 움직임에 차별점을 두어 테스트하는 것은 제가 구현한 의존성 주입 설계 구성으로는 아직 부족하다고 느꼈습니다. 물론 전자의 테스트만으로도 검증에 문제가 없으므로 후자가 꼭 필요한 테스트는 아니지만, 이런 생각까지 해볼 수 있게 되어 감사합니다.
> 
> 앞으로 이러한 점들을 기반으로 향후 미션에서 더 넓게 생각하고 적용할 수 있는 사람이 되겠습니다. 1주차 미션 리뷰 도와주셔서 감사합니다.

@Hevton 
팡테 안녕하세요~
잘 구현해주셔서 반드시 수정이 필요한 건 아니었고 참고해볼만한 내용으로 남겨둔거라서 앞으로 차차 미션 진행해나가면서 고민해보셔도 좋을 것 같습니다ㅎㅎ

이번 미션 진행하시느라 고생 많으셨습니다🙇‍♂️
다음 미션도 화이팅입니다!!💪
